### DISCUSSION TITLE
Significant performance degradation on Payload 3

### DISCUSSION THREAD

**Discussion by immotus** [payloadcms/payload#9403] [General]
I followed the migration guide from Payload v2 to v3 for a setup on MongoDB and Cloud Storage (GCP). Most of the migration went really without significant hiccups. However, the performance of the new Next-based Payload is pathetically slow. For example, It takes about seconds 2-3 seconds to switch between Edit mode to Versions, to API.

Look at these times, on a fairly small project.

```bash
> cross-env NODE_OPTIONS=--no-deprecation next dev

   ▲ Next.js 15.0.3
   - Local:        http://localhost:3000
   - Environments: .env
   - Experiments (use with caution):
     · turbo

 ✓ Starting...
It looks like you're trying to use TypeScript but do not have the required package(s) installed.
Installing dependencies

If you are not trying to use TypeScript, please remove the tsconfig.json file from your package root (and any TypeScript files in your app and pages directories).


Installing devDependencies (npm):
- @types/react


up to date, audited 795 packages in 3s

167 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

 ✓ Ready in 5.5s
 ○ Compiling /admin/[[...segments]] ...
 ✓ Compiled /admin/[[...segments]] in 21.2s (4596 modules)
[22:57:50] WARN: No email adapter provided. Email will be written to console. More info at https://payloadcms.com/docs/email/overview.
 GET /admin 307 in 34142ms
 GET /admin 307 in 813ms
 GET /admin/login 200 in 566ms
 ○ Compiling /api/[...slug] ...
 ✓ Compiled /api/[...slug] in 3.3s (4767 modules)
 GET /api/payload-preferences/nav 200 in 6039ms
 GET /api/users/me 200 in 6092ms
 POST /api/users/login 200 in 1911ms
 GET /api/payload-preferences/locale 200 in 505ms
 GET /admin 200 in 781ms
 GET /admin/collections/pages 200 in 1501ms
 POST /admin/collections/pages 200 in 297ms
 POST /admin/collections/pages 200 in 253ms
 POST /admin/collections/pages 200 in 245ms
 GET /admin/collections/pages?limit=10 200 in 1027ms
 GET /admin/collections/pages?limit=10 200 in 1733ms
 GET /admin/collections/pages/66eada6b37de69690684b32d 200 in 2207ms
 GET /api/payload-preferences/collection-pages-66eada6b37de69690684b32d 200 in 510ms
 GET /api/media/662265046f9e2dc53ac5eeb7?depth=0 200 in 17ms
 GET /api/media/662a6f316c2946b69436aa35?depth=0 200 in 18ms
 GET /api/media/66d2ba159ea593f0d43755e7?depth=0 200 in 24ms
 GET /api/media/662a6f316c2946b69436aa35?depth=0 200 in 489ms
 GET /api/media/662265046f9e2dc53ac5eeb7?depth=0 200 in 493ms
 GET /api/media/6622663a9b679b63a755fc2e?depth=0 200 in 496ms
 GET /api/media/6622691462305b5ae2d8b0c8?depth=0 200 in 498ms
 GET /api/media/66d2ba159ea593f0d43755e7?depth=0 200 in 1713ms
 GET /api/media/file/Scarlet%20robin-400x400.webp 200 in 1907ms
 GET /api/media/file/cockatoo-400x400.webp 200 in 1957ms
 GET /api/media/file/Spiral%20Aloe-400x400.webp 200 in 2397ms
 GET /api/media/file/Birds-400x400.webp 200 in 1764ms
 GET /api/media/file/pink-400x400.webp 200 in 4044ms
 GET /admin/collections/pages/66eada6b37de69690684b32d/versions 200 in 3455ms
 GET /admin/collections/pages/66eada6b37de69690684b32d/versions?limit=10 200 in 3442ms
```

It is about 50-100 times slower than v2.

Just look at the start-up time and the initial load of the `/admin` route:

```
GET /admin 307 in 34142ms
````

User info in 6 seconds?

```
GET /api/users/me 200 in 6092ms
```

I conducted the test with the CLI-produced `website` example on MongoDB, and it is just as slow. Was this whole idea to move to Next.js not worth it?

**Comment by andershermansen** [comment]
You are running it in dev mode. For best performance you should build it and run it in production mode.

Instead of
```sh
pnpm dev
````
run
```sh
pnpm build
pnpm start
```

Another thing to think about regarding performance is the roundtrip to the database. Where is the code running vs where is the database running. If they are far apart on slow internet, it will go slow. If they are near with fast connection it will be faster.

**Reply by immotus** [reply]
The production build is still failing, I'm still working on fixes. So I haven't tested that yet. But with all other things being equal, i.e. location of the database etc., the dev mode on v2 was never so laggy.

**Reply by immotus** [reply]
Just to be sure, as a matter of test, I did a clean install of the Payload Website App using `npx create-payload-app`, seeded the database, and simply switching between Edit, Version, and API modes in the admin panel takes 2-3 seconds, whilst it was nearly instantaneous in v2.

**Reply by andershermansen** [reply]
I'm not sure what your issue is. I was very unhappy with v3 speed using serverless (vercel) hosting in the early betas, it felt very sluggish. Have not tested that type of hosting again in quite some time. I switched to VPS hosting with coolify and using a database (mongo atlas) in location near the VPS and of course production build. After that I am very happy with the performance. Probably the payload team will work more on performance now that v3 is out 🤞

**Reply by andershermansen** [reply]
Note that if you look at the PR's the team is working on right now, there are a lot of things going on with performance. It looks very promising. So it seems the team is working hard to make sure it's as fast as possible for everyone. 3.0 has just been out for a week.

**Reply by immotus** [reply]
> Note that if you look at the PR's the team is working on right now, there are a lot of things going on with performance. It looks very promising. So it seems the team is working hard to make sure it's as fast as possible for everyone. 3.0 has just been out for a week.

I actually did try to look for open issues relating to performance, but couldn't find anything resembling the issues I'm experiencing. This is why I have started this conversation.

**Reply by immotus** [reply]
I'm looking forward to improvements in this space because the performance, especially compared with the previous version, is appalling.

**Reply by andershermansen** [reply]
@immotus Out of curiosity. Are you using mongodb or postgres? How are you hosting the database?
And how are you hosting payload itself?

**Reply by immotus** [reply]
> @immotus Out of curiosity. Are you using mongodb or postgres? How are you hosting the database? And how are you hosting payload itself?

The working version of Payload 2 is hosted in a GCP Cloud Run express app with GCP-hosted MongoDB. With Payload 3, I'm still struggling with the migration, so it is all local.

**Reply by andershermansen** [reply]
Ok, so with both database and code local the latency should not be an issue. Also serverless is not an issue.

Are you running with in dev mode `pnpm dev` or in production mode `pnpm build && pnpm start` ?
Also, do you have complex/big collection definitions? Some of the optimizations I see in progress will have big effects on large definitions.

**Reply by immotus** [reply]
It is a production build with a very simple definition of a Global. Same as on v2.

**Comment by immotus** [comment]
@jmikrut is there any sensible explanation why a simple Global retrieval in Payload 3 (local production build with a local MongoDB) works three times slower (~1000 ms) than in Payload 2 deployed in the Cloud with a remote MongoDB (~180-200 ms). With all/most other things being equal, it appears to me that Next.js is just so much slower than the old way Payload was run.
